From b3cc395eddfc0583bba0e2d302230ebe8770599d Mon Sep 17 00:00:00 2001
From: Mattias Ellert <mattias.ellert@physics.uu.se>
Date: Mon, 2 Jan 2023 12:51:26 +0100
Subject: [PATCH 2/6] Fix warning about possible use after free

Fix also a warning about an unused variable. This piece of code can probably
go away, but let's just fix the warning for the time being.
---
 src/sslutils/sslutils.c | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/src/sslutils/sslutils.c b/src/sslutils/sslutils.c
index 6bac77e..8b0f4c4 100644
--- a/src/sslutils/sslutils.c
+++ b/src/sslutils/sslutils.c
@@ -455,7 +455,7 @@ ERR_load_prxyerr_strings(
 #else
     char *                              randfile;
 #endif
-#if SSLEAY_VERSION_NUMBER >=  0x0090581fL
+#if SSLEAY_VERSION_NUMBER >=  0x0090581fL && !defined(OPENSSL_NO_EGD)
     char *                              egd_path;
 #endif
     char                                buffer[200];
@@ -2703,8 +2703,14 @@ proxy_get_filenames(
                       }
 
                     }
-                    else
-                      strcpy(default_user_cert, certname);
+                    else {
+                      default_user_cert = strndup(certname, strlen(certname));
+
+                      if (!default_user_cert) {
+                        PRXYerr(PRXYERR_F_INIT_CRED, PRXYERR_R_OUT_OF_MEMORY);
+                        goto err;
+                      }
+                    }
 
                     default_user_key = strndup(default_user_cert, strlen(default_user_cert));
 
-- 
2.38.1

From 77020a5574bfdbdcdebe03a3ee5d3d3f99563c03 Mon Sep 17 00:00:00 2001
From: Mattias Ellert <mattias.ellert@physics.uu.se>
Date: Mon, 2 Jan 2023 12:53:26 +0100
Subject: [PATCH 3/6] Fix doxygen warning

About a documented return type for a function that does not return anything.
---
 src/api/ccapi/voms_apic.h | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/src/api/ccapi/voms_apic.h b/src/api/ccapi/voms_apic.h
index 8807803..a3551f4 100644
--- a/src/api/ccapi/voms_apic.h
+++ b/src/api/ccapi/voms_apic.h
@@ -205,8 +205,7 @@ extern struct contactdata **VOMS_FindByVO(struct vomsdata *vd, char *vo,
 
 
 extern void VOMS_DeleteContacts(struct contactdata **list); /*!< Frees a contactdata vector.
-							      \param list The vector to free.
-							      \return NONE */
+							      \param list The vector to free.*/
 
 extern struct vomsdata *VOMS_Init(char *voms, char *cert); /*!< Initializes a vomsdata structure for use by the other functions.
 						       N.B: This is the ONLY way to correctly initialize a vomsdata structure. It
-- 
2.38.1

From bda11dca2561d937f0452d710a0c6755e9b92c6d Mon Sep 17 00:00:00 2001
From: Mattias Ellert <mattias.ellert@physics.uu.se>
Date: Mon, 2 Jan 2023 13:08:28 +0100
Subject: [PATCH 4/6] Fix warning about possible string truncation

This is a false positive, since the source is an 8-byte hash and
is copied into an 8-byte substring. memcpy is a better fit anyway.
---
 src/sslutils/evaluate.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/sslutils/evaluate.c b/src/sslutils/evaluate.c
index 09b8ba4..9c03fdc 100644
--- a/src/sslutils/evaluate.c
+++ b/src/sslutils/evaluate.c
@@ -353,8 +353,8 @@ void PRIVATE read_pathrestriction(STACK_OF(X509) *chain, char *path,
     hash = gethash(cert, hashed);
 
     /* Determine file names */
-    strncpy(signing + 1, hash, 8);
-    strncpy(namespace + 1, hash, 8);
+    memcpy(signing + 1, hash, 8);
+    memcpy(namespace + 1, hash, 8);
 
     file = open_from_dir(path, signing);
     if (file) {
-- 
2.38.1

